### PR TITLE
Wr2

### PR DESCRIPTION
**Pull request by glennw**

### CODE REVIEW

**Comment by glennw** [comment]
r? @pcwalton 

I'm very sorry to drop such a large patch on you - let's go through it in person tomorrow?

**Comment by metajack** [comment]
http://giphy.com/gifs/kgSleqI7SguTS/html5

File: res/ps_border.fs.glsl [hunk 1]
```diff
@@ -58,10 +58,28 @@ vec4 draw_dotted_edge() {
...
+vec2 get_dashed_nudge_factor(vec2 dash_size, bool is_corner) {
+  if (is_corner) {
+    return vec2(0.0, 0.0);
+  }
+
+  bool xAxisFudge = vBorders.z > vBorders.w;
+  if (xAxisFudge) {
+    return vec2(dash_size.x / 2.0, 0);
+  } else {
+    return vec2(0.0, dash_size.y / 2.0);
```
Comment (pcwalton) [inline res/ps_border.fs.glsl]: nit: `else` is redundant

File: res/ps_border.fs.glsl [hunk 2]
```diff
@@ -58,10 +58,28 @@ vec4 draw_dotted_edge() {
...
+  }
+}
+
+vec4 draw_dashed_edge(bool is_corner) {
   // Everything here should be in device pixels.
   // We want the dot to be roughly the size of the whole border spacing
-  float dash_interval = min(vBorders.w, vBorders.z) * 3;
+  // 5.5 here isn't a magic number, it's just what mostly looks like FF/Chrome
+  float dash_interval = min(vBorders.w, vBorders.z) * 5.5;
```
Comment (pcwalton) [inline res/ps_border.fs.glsl]: Would be nice to eventually look in the Gecko/WebKit source and see precisely what they do

File: res/ps_border.fs.glsl [hunk 3]
```diff
@@ -49,6 +58,33 @@ vec4 draw_dotted_edge() {
...
+  vec4 target_rect = vec4(target_rect_loc, dash_size);
+
+  vec4 white = vec4(1.0, 1.0, 1.0, 1.0);
+  vec3 black = vec3(0.0, 0.0, 0.0);
+  vec4 target_colored_rect = drawRect(position, target_rect, black);
+
+  return mix(white, target_colored_rect, target_colored_rect.a);
+
+  //vec4 white = vec4(1.0, 1.0, 1.0, 1.0);
+  //return white;
```
Comment (pcwalton) [inline res/ps_border.fs.glsl]: Remove this commented out code?

File: res/ps_border.fs.glsl [hunk 4]
```diff
@@ -68,6 +104,25 @@ void draw_dotted_border(void) {
...
+    case PST_BOTTOM_RIGHT:
+      // TODO: Fix for corners with a border-radius
+      oFragColor = draw_dashed_edge();
+      break;
+    case PST_BOTTOM:
+    case PST_TOP:
+    case PST_LEFT:
+    case PST_RIGHT:
+      oFragColor = draw_dashed_edge();
+      break;
```
Comment (pcwalton) [inline res/ps_border.fs.glsl]: We should have an issue to try to remove this branching in the shader somehow.

File: res/ps_border.vs.glsl [hunk 5]
```diff
@@ -123,6 +123,6 @@ void main(void) {
     // These are in device space
     vPos = clamped_pos;
     vBorders = vec4(border.local_rect.x, border.local_rect.y,
-                    border.local_rect.x + border.local_rect.z,
-                    border.local_rect.y + border.local_rect.w) * uDevicePixelRatio;
+                    border.local_rect.z,
+                    border.local_rect.w) * uDevicePixelRatio;
```
Comment (pcwalton) [inline res/ps_border.vs.glsl]: Isn't this just equivalent to `vBorders = border.local_rect * uDevicePixelRatio;` ?

File: res/ps_angle_gradient.fs.glsl [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    return clamp((x - lo) / (hi - lo), 0.0, 1.0);
+}
+
+void main(void) {
+    float angle = atan(-vEndPoint.y + vStartPoint.y,
+                        vEndPoint.x - vStartPoint.x);
+    float sa = sin(angle);
+    float ca = cos(angle);
+
+	float sx = vStartPoint.x * ca - vStartPoint.y * sa;
```
Comment (pcwalton) [inline res/ps_angle_gradient.fs.glsl]: nit: weird indentation here

File: res/ps_box_shadow.vs.glsl [hunk 7]
```diff
@@ -0,0 +1,51 @@
+#line 1
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+struct BoxShadow {
+	PrimitiveInfo info;
+	vec4 local_rect;
+	vec4 color;
+    vec4 border_radii_blur_radius_inverted;
```
Comment (pcwalton) [inline res/ps_box_shadow.vs.glsl]: nit: weird indentation here

File: res/ps_gradient.vs.glsl [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#define DIR_HORIZONTAL      uint(0)
+#define DIR_VERTICAL        uint(1)
+
+struct Gradient {
+	PrimitiveInfo info;
+    vec4 local_rect;
```
Comment (pcwalton) [inline res/ps_gradient.vs.glsl]: nit: weird indentation here

File: res/shared.glsl [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+//======================================================================================
+// Vertex shader attributes and uniforms
+//======================================================================================
+#ifdef WR_VERTEX_SHADER
+    #define varying out
+
+    // Uniform inputs
+	uniform mat4 uTransform;       // Orthographic projection
+    uniform float uDevicePixelRatio;
```
Comment (pcwalton) [inline res/shared.glsl]: nit: weird indentation

File: res/shared_other.glsl [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+// Vertex shader attributes and uniforms
+//======================================================================================
+#ifdef WR_VERTEX_SHADER
+    in vec4 aColorTexCoordRectTop;
+    in vec4 aColorRectTL;
+
+    // box-shadow
+    in vec4 aBorderPosition;
+	in vec4 aBorderRadii;
+	in float aBlurRadius;
```
Comment (pcwalton) [inline res/shared_other.glsl]: nit: weird indentation

File: res/ps_border.fs.glsl [hunk 11]
```diff
@@ -9,47 +9,24 @@ vec4 drawCircle(vec2 aPixel, vec2 aDesiredPos, float aRadius, vec3 aColor) {
...
-  float radius = floor(border_spacing / spacing_fudge);
+  float radius = floor(border_spacing / 2.0);
   float diameter = radius * 2.0;
-  float circleSpacing = diameter * 2.0;
+  // The amount of space between dots. 2.2 was chosen because it looks kind of
+  // like firefox.
+  float circleSpacing = diameter * 2.2;
 
-  vec2 size = vec2(vBorders.z - vBorders.x, vBorders.w - vBorders.y);
+  vec2 size = vec2(vBorders.z, vBorders.w);
```
Comment (pcwalton) [inline res/ps_border.fs.glsl]: nit: could just be `vec2 size = vBorders.zw;`

File: src/internal_types.rs [hunk 12]
```diff
@@ -694,12 +690,12 @@ impl BoxShadowRasterOp {
             None
         }
     }
-}
+}*/
```
Comment (pcwalton) [inline src/internal_types.rs]: Maybe just remove this?

File: src/internal_types.rs [hunk 13]
```diff
@@ -723,7 +719,7 @@ impl GlyphKey {
 
 #[derive(Clone, Debug, Hash, Eq, PartialEq)]
 pub enum RasterItem {
-    BoxShadow(BoxShadowRasterOp),
+    _BoxShadow(BoxShadowRasterOp),
 }
```
Comment (pcwalton) [inline src/internal_types.rs]: Maybe just remove this enum entirely?

File: src/resource_list.rs [hunk 14]
```diff
@@ -27,7 +26,7 @@ impl ResourceList {
             required_glyphs: HashMap::with_hasher(Default::default()),
             required_images: HashSet::with_hasher(Default::default()),
             required_rasters: HashSet::with_hasher(Default::default()),
-            device_pixel_ratio: device_pixel_ratio,
+            _device_pixel_ratio: device_pixel_ratio,
```
Comment (pcwalton) [inline src/resource_list.rs]: Maybe remove this?

File: src/resource_list.rs [hunk 15]
```diff
@@ -69,7 +69,7 @@ impl ResourceList {
                                                                   self.device_pixel_ratio) {
             self.required_rasters.insert(RasterItem::BoxShadow(raster_item));
         }
-    }
+    }*/
```
Comment (pcwalton) [inline src/resource_list.rs]: Just remove this?

File: src/util.rs [hunk 16]
```diff
@@ -200,59 +134,78 @@ pub fn rect_is_empty<N:PartialEq + Zero>(rect: &Rect<N>) -> bool {
...
-    rect.size.width > 0.0 && rect.size.height > 0.0
-}
+/*
+#[inline(always)]
+pub fn rect_contains_rect(rect: &Rect<Au>, other: &Rect<Au>) -> bool {
+    rect.origin.x <= other.origin.x &&
+    rect.origin.y <= other.origin.y &&
+    rect.max_x() >= other.max_x() &&
+    rect.max_y() >= other.max_y()
+}*/
```
Comment (pcwalton) [inline src/util.rs]: Just remove this?

File: src/util.rs [hunk 17]
```diff
@@ -119,78 +117,14 @@ impl RectHelpers for Rect<f32> {
...
-        }
-    }
-    fn accumulate(values: &[ColorF; 4]) -> ColorF {
-        ColorF {
-            r: values[0].r + values[1].r + values[2].r + values[3].r,
-            g: values[0].g + values[1].g + values[2].g + values[3].g,
-            b: values[0].b + values[1].b + values[2].b + values[3].b,
-            a: values[0].a + values[1].a + values[2].a + values[3].a,
-        }
+    fn is_well_formed_and_nonempty(&self) -> bool {
```
Comment (pcwalton) [inline src/util.rs]: nit: I'd copy the comment over instead of deleting it

File: src/util.rs [hunk 18]
```diff
@@ -3,10 +3,9 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 use euclid::{Matrix4D, Point2D, Point4D, Rect, Size2D};
-use internal_types::{RectColors};
+use internal_types::{DevicePixel};
```
Comment (pcwalton) [inline src/util.rs]: nit: no `{}`

File: src/tiling.rs [hunk 19]
```diff
@@ -0,0 +1,2147 @@
...
+                     clip: Option<Box<Clip>>,
+                     details: PrimitiveDetails) {
+        let current_layer = *self.layer_stack.last().unwrap();
+        let StackingContextIndex(layer_index) = current_layer;
+        let layer = &mut self.layers[layer_index as usize];
+
+        let prim = Primitive {
+            rect: *rect,
+            xf_rect: None,
+
```
Comment (pcwalton) [inline src/tiling.rs]: Just remove the pass count?

Comment (pcwalton) [inline src/tiling.rs]: I thought we did support angle gradients?

File: src/tiling.rs [hunk 20]
```diff
@@ -0,0 +1,2147 @@
...
+        let first_opaque_cover_index = self.prim_indices.iter().position(|i| {
+            let PrimitiveIndex(pi) = *i;
+            let prim = &layer.primitives[pi];
+            prim.is_opaque() &&
+               rect_contains_rect(&prim.xf_rect.as_ref().unwrap().bounding_rect, screen_rect)
+        });
+        if let Some(first_opaque_cover_index) = first_opaque_cover_index {
+            self.prim_indices.truncate(first_opaque_cover_index);
+        }
+*/
```
Comment (pcwalton) [inline src/tiling.rs]: Just remove this with `TODO: Intra-layer occlusion.`

File: src/tiling.rs [hunk 21]
```diff
@@ -0,0 +1,2147 @@
...
+}
+
+impl StackingContext {
+    fn build_resource_list(&self,
+                           resource_list: &mut ResourceList,
+                           //index_buffer: &Vec<PrimitiveIndex>,
+                           auxiliary_lists: &AuxiliaryLists) {
+//        for prim_index in index_buffer {
+//            let PrimitiveIndex(prim_index) = *prim_index;
+//            let prim = &self.primitives[prim_index];
```
Comment (pcwalton) [inline src/tiling.rs]: Remove this?

File: src/tiling.rs [hunk 22]
```diff
@@ -0,0 +1,2147 @@
...
+    scroll_layer_id: ScrollLayerId,
+    opacity: f32,
+    transform: Matrix4D<f32>,
+    xf_rect: Option<TransformedRect>,
+}
+
+impl StackingContext {
+    fn build_resource_list(&self,
+                           resource_list: &mut ResourceList,
+                           //index_buffer: &Vec<PrimitiveIndex>,
```
Comment (pcwalton) [inline src/tiling.rs]: Remove this?

File: src/tiling.rs [hunk 23]
```diff
@@ -0,0 +1,2147 @@
...
+            src_rects: unsafe { mem::uninitialized() },
+            blend_info: [ 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0 ],
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Hash, Clone)]
+pub struct CompositeBatchKey {
+    pub shader: CompositeShader,
+    //pub samplers: [TextureId; MAX_PRIMS_PER_COMPOSITE],
```
Comment (pcwalton) [inline src/tiling.rs]: Remove this

File: src/tiling.rs [hunk 24]
```diff
@@ -0,0 +1,2147 @@
...
+                                                              0.0,
+                                                              1.0)),
+                    transform.transform_point4d(&Point4D::new(rect.top_right().x,
+                                                              rect.top_right().y,
+                                                              0.0,
+                                                              1.0)),
+                ];
+
+                let mut screen_min: Point2D<f32> = Point2D::new( 10000000.0,  10000000.0);
+                let mut screen_max: Point2D<f32> = Point2D::new(-10000000.0, -10000000.0);
```
Comment (pcwalton) [inline src/tiling.rs]: Move these to statics

File: src/texture_cache.rs [hunk 25]
```diff
@@ -107,6 +103,12 @@ impl TexturePage {
         page
     }
 
+/*
+    pub fn size(&self) -> u32 {
+        self.texture_size
+    }
+*/
```
Comment (pcwalton) [inline src/texture_cache.rs]: Just remove this

File: src/resource_cache.rs [hunk 26]
```diff
@@ -473,9 +439,9 @@ fn run_raster_jobs(thread_pool: &mut scoped_threadpool::Pool,
                                                         device_pixel_ratio,
                                                         enable_aa);
                 });
-            });
+            //});
         }
-    });
+    //});
```
Comment (pcwalton) [inline src/resource_cache.rs]: Remove this and say "TODO: Run raster jobs in parallel again"

File: src/resource_cache.rs [hunk 27]
```diff
@@ -442,7 +408,7 @@ impl ResourceCache {
     }
 }
 
-fn run_raster_jobs(thread_pool: &mut scoped_threadpool::Pool,
+fn run_raster_jobs(//thread_pool: &mut scoped_threadpool::Pool,
```
Comment (pcwalton) [inline src/resource_cache.rs]: Just remove this parameter

File: src/resource_cache.rs [hunk 28]
```diff
@@ -420,20 +386,20 @@ impl ResourceCache {
...
+/*
     #[inline]
     pub fn get_webgl_texture(&self, context_id: &WebGLContextId) -> TextureId {
         self.webgl_textures.get(context_id).unwrap().clone()
     }
-
-    pub fn device_pixel_ratio(&self) -> f32 {
-        self.device_pixel_ratio
-    }
+*/
```
Comment (pcwalton) [inline src/resource_cache.rs]: Just remove these

File: src/resource_cache.rs [hunk 29]
```diff
@@ -420,20 +386,20 @@ impl ResourceCache {
         self.texture_cache.get(image_info.texture_cache_id)
     }
 
+/*
     #[inline]
     pub fn get_raster(&self, raster_item: &RasterItem, frame_id: FrameId) -> &TextureCacheItem {
         let image_id = self.cached_rasters.get(raster_item, frame_id);
         self.texture_cache.get(*image_id)
     }
+*/
```
Comment (pcwalton) [inline src/resource_cache.rs]: Just remove this

File: src/resource_cache.rs [hunk 30]
```diff
@@ -130,15 +130,13 @@ pub struct ResourceCache {
...
-    dummy_mask_image_id: TextureCacheItemId,
+    //white_image_id: TextureCacheItemId,
 }
 
 impl ResourceCache {
     pub fn new(thread_pool: &mut scoped_threadpool::Pool,
                texture_cache: TextureCache,
-               white_image_id: TextureCacheItemId,
-               dummy_mask_image_id: TextureCacheItemId,
+               _white_image_id: TextureCacheItemId,
```
Comment (pcwalton) [inline src/resource_cache.rs]: Do we not need this anymore? If so, just remove

File: src/resource_cache.rs [hunk 31]
```diff
@@ -130,15 +130,13 @@ pub struct ResourceCache {
 
     pending_raster_jobs: Vec<GlyphRasterJob>,
 
-    white_image_id: TextureCacheItemId,
-    dummy_mask_image_id: TextureCacheItemId,
+    //white_image_id: TextureCacheItemId,
```
Comment (pcwalton) [inline src/resource_cache.rs]: Just remove

File: src/renderer.rs [hunk 32]
```diff
@@ -1153,728 +1175,408 @@ impl Renderer {
...
+                            tile_y0,
+                            c,//&ColorF::new(0.0, 0.0, 0.0, 1.0),
+                            tile_x1,
+                            tile_y1,
+                            c);//&ColorF::new(0.0, 0.0, 0.0, 1.0));
+        if label.len() > 0 {
+            self.debug.add_text((tile_x0.0 as f32 + tile_x1.0 as f32) * 0.5,
+                                (tile_y0.0 as f32 + tile_y1.0 as f32) * 0.5,
+                                label,
+                                c);//&ColorF::new(0.0, 0.0, 0.0, 1.0));
```
Comment (pcwalton) [inline src/renderer.rs]: Remove commented out code here and above

File: src/renderer.rs [hunk 33]
```diff
@@ -412,26 +577,32 @@ impl Renderer {
...
         profile_timers.cpu_time.profile(|| {
             self.device.begin_frame();
 
             gl::disable(gl::SCISSOR_TEST);
-            gl::clear_color(1.0, 1.0, 1.0, 0.0);
+            //gl::clear_color(1.0, 1.0, 1.0, 0.0);
+            //gl::clear(gl::COLOR_BUFFER_BIT | gl::DEPTH_BUFFER_BIT | gl::STENCIL_BUFFER_BIT);
 
-            self.update_shaders();
+            //self.update_shaders();
```
Comment (pcwalton) [inline src/renderer.rs]: Remove commented out stuff here

File: src/render_backend.rs [hunk 34]
```diff
@@ -196,27 +199,37 @@ impl RenderBackend {
...
                             let frame = profile_counters.total_time.profile(|| {
                                 if self.frame.scroll(delta, cursor, move_phase) {
+                                    self.build_scene();
                                     Some(self.render())
                                 } else {
                                     None
                                 }
                             });
+//                            hprof::end_frame();
+                            //hprof::profiler().print_timing();
```
Comment (pcwalton) [inline src/render_backend.rs]: Remove commented out code

File: src/render_backend.rs [hunk 35]
```diff
@@ -196,27 +199,37 @@ impl RenderBackend {
                                 self.build_scene();
                                 self.render()
                             });
+//                            hprof::end_frame();
+                            //hprof::profiler().print_timing();
```
Comment (pcwalton) [inline src/render_backend.rs]: Remove commented out code

File: src/kdtree.rs [hunk 36]
```diff
@@ -0,0 +1,4 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
```
Comment (pcwalton) [inline src/kdtree.rs]: Delete this file?

File: src/layer.rs [hunk 37]
```diff
@@ -65,35 +56,12 @@ impl Layer {
...
-        self.aabb_tree.insert(rect,
-                              draw_list_group_id,
-                              draw_list_id,
-                              item_index);
-    }
-
     pub fn finalize(&mut self, scrolling: &ScrollingState) {
         self.scrolling = *scrolling;
-        self.aabb_tree.finalize();
+        //self.aabb_tree.finalize();
```
Comment (pcwalton) [inline src/layer.rs]: Just delete this line

File: src/layer.rs [hunk 38]
```diff
@@ -105,6 +73,7 @@ impl Layer {
     pub fn print(&self) {
         self.aabb_tree.print(NodeIndex(0), 0);
     }
+*/
```
Comment (pcwalton) [inline src/layer.rs]: Remove this commented out code

File: src/lib.rs [hunk 39]
```diff
@@ -99,6 +101,7 @@ extern crate app_units;
 extern crate euclid;
 extern crate fnv;
 extern crate gleam;
+//extern crate hprof;
```
Comment (pcwalton) [inline src/lib.rs]: Remove this

File: src/quadtree.rs [hunk 40]
```diff
@@ -0,0 +1,222 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (pcwalton) [inline src/quadtree.rs]: Just delete this file

**Comment by pcwalton** [comment]
Here we go!

**Comment by Tankenstein** [comment]
🎉

**Pull request merged**